### ISSUE TITLE
Android 7.1.2 (LineageOS Guava) does not display colors, all objects are rendered in black

### ISSUE THREAD

**Issue by teodorg** [aframevr/aframe#3523]
**Description:**

On Android 7.1.2 (LineageOS Guava) all objects are displayed in solid black color. Tested with different browsers (Mozilla and Firefox).
Tested on a Motorolla Moto G3. The same device works fine with Android 6 and the problem appears
once it switches to the latest updates of LineageOS 7.1.2.

The problem does not exist on Android 5 and 6.

- A-Frame Version: all
- Platform / Device: Android 7.1.2 (LineageOS Guava) / Motorolla Moto G3
- Reproducible Code Snippet or URL:

**Comment by gabrielgrant** [comment]
I don't think this is specific LineageOS -- I'm seeing this using Chrome 66.0.3359.158 with LG's stock Android 7.0  (version M21010v; January 1, 2018 security update) with on both my LG Aristo (LG-M210) and my LG Fiesta (LG L63BL)

**Comment by dmarcos** [comment]
There must be something specific
to those devices. Do you see any errors on the browser console?

**Comment by teodorg** [comment]
@gabrielgrant It is not browser specific as well, because I tried with the default Android browser, Firefox and a beta version of Chrome.

@dmarcos It is not specific to the device, because I have two completely identical  Motorolla Moto G3 handsets (well, the only difference between them is their color - white and black). One is running LineageOS 7.1.2 and one is running stock Android 6.0.1. The one, equipped with 6.0.1 version renders correctly. The "blacked objects" issue appears only on the LineageOS-equipped handset. Yet, as far as I remember, first my code ran fine on that handset as well and the problem appeared after one of the updates.

**Comment by gabrielgrant** [comment]
@dmarcos The only errors I see are are two instances of `Failed to recalculate device parameters.` being logged out
![screenshot from 2018-05-24 01 50 16](https://user-images.githubusercontent.com/493968/40474976-3baf67ec-5ef5-11e8-9b25-e811f018b64d.png)

This is coming from here:
![screenshot from 2018-05-24 01 52 30](https://user-images.githubusercontent.com/493968/40475006-49ba6670-5ef5-11e8-90c7-db606ec31519.png)


Full stack if you're interested: 
![screenshot from 2018-05-24 01 50 58](https://user-images.githubusercontent.com/493968/40475032-575ea728-5ef5-11e8-82e0-2b1ac4961975.png)

This seems to be (trying) to load some sort of device parameters from https://dpdb.webvr.rocks/dpdb.json and isn't finding an entry for this device (my device's model number -- LG-M210 -- does not appear in that database). I'm not sure what these parameters are used for, but if this is actually the problem, it seems like this might explain why it is device-specific (though not, afaict, why it would be Android-version specific: the User-Agent string only seems to be matched against device ID). 

![screenshot from 2018-05-24 02 09 40](https://user-images.githubusercontent.com/493968/40476324-9e734486-5ef8-11e8-84c5-61b8f937a97c.png)


This DB's info seems to mostly be about display size, though. AFAICS the canvas, at least, seems to be getting set up correctly, so it's totally possible this is a red herring. OTOH it's also possible the black models are coming from some problem with a texture buffer being improperly allocated (perhaps because of this exact missing resolution info).

In either case, I'm a bit out of my depth. Looks like both you and @ngokevin are in the bay area. I'm in SF too and have several of these devices for testing. I'd be happy to get you one of them, or can just supply you with more info if that's easier.

**Comment by dmarcos** [comment]
It sounds like a problem at the drivers level of your OS + device combo. Something might not be working at the shader level since it works for other OSs versions. Just curious. Is this THREE example rendering correctly: https://threejs.org/docs/scenes/material-browser.html#MeshStandardMaterial

**Comment by dmarcos** [comment]
It seems the problem is deep in the guts of the OS / Drivers / Browser / WebGL
implementation. We cannot do much on the A-Frame side. I keep this issue open as a reference and to collect further info.

**Comment by gabrielgrant** [comment]
Hmmm...it certainly isn't a problem with all WebGL texturing, though -- https://fltr.world/ works fine

**Comment by dmarcos** [comment]
If I have to bet is a problem with the shaders. The shader branch of the MeshStandardMaterial is hitting some bug in the OS / Device / Driver combo you have.

**Comment by dmarcos** [comment]
@gabrielgrant Would you mind opening an issue on THREE.js too?

**Comment by gabrielgrant** [comment]
so to clarify: you don't think that device parameter DB issue is related, right?

**Comment by gabrielgrant** [comment]
@dmarcos Yup, was just in the process of searching their issue tracker for anything related before doing so :)

**Comment by dmarcos** [comment]
No, I don’t the DB issue is related since also reproduces on THREE.js

**Comment by gabrielgrant** [comment]
Makes sense, just wanted to confirm

**Comment by gabrielgrant** [comment]
Filed here: mrdoob/three.js#14137

**Comment by gabrielgrant** [comment]
@machenmusik thanks, should have thought of using other non-flat shaders. Tried a few and posted the results here: https://github.com/mrdoob/three.js/issues/14137#issuecomment-391880854

**Comment by machenmusik** [comment]
I think only standard and flat are built in at the moment.  But @dmarcos @ngokevin @donmccurdy is there any particular reason?  Request for Lambert sounds reasonable, for example... And THREE includes all of them either way.

**Comment by dmarcos** [comment]
As don said there has not been much demand for lambert. It would one more code path to maintain.

**Comment by gabrielgrant** [comment]
Nope: 
`<a-entity geometry="primitive: box" material="flatShading: true; color: blue"></a-entity>` just renders as black

`<a-entity geometry="primitive: box" material="shader: flat; color: blue"></a-entity>` renders as blue, but (obviously) flat

**Comment by dmarcos** [comment]
It's a problem with the shader path for the standard material in THREE and your particular combo of device / OS / driver / firmware: 99% sure.

**Comment by donmccurdy** [comment]
Yeah, I was curious about `flatShading` because it compiles away one difference between standard and lambert but apparently that wasn't it..

**Comment by gabrielgrant** [comment]
@dmarcos yes, that certainly sounds like the issue is a problem with the shader using something unsupported. Though if it is, in fact, the same problem as the previously-closed related issue, it doesn't sound particularly specific to any given device: we have repros on at least 6 different devices from 4 manufacturers, and @EerikKivistik [says they "are getting a few hundred reports a week from our users"](https://github.com/mrdoob/three.js/issues/12679#issuecomment-380071840)

My guess is the issue is something to do with the Qualcomm Snapdragon 425/Adreno 308 GPU -- I believe that's the (extremely common) low-/mid-range chip used in most (all?) of the devices reporting this issue (as well as most other common low-/mid-range Android devices available in the market these days)

**Comment by gabrielgrant** [comment]
@teodorg would you be able to confirm that the two test cases we've isolated in [the three.js issue](https://github.com/mrdoob/three.js/issues/14137) are reproducible on your Moto G3?

(note: the second issue causes the phone to crash completely and hard-restart, so make sure you don't have anything important open that you'll lose, and I'd recommend testing it in an incognito tab to avoid getting stuck in [the same restart loop I did](https://github.com/mrdoob/three.js/issues/14137#issuecomment-396348889))

https://bugs.chromium.org/p/chromium/issues/detail?id=851870
https://bugs.chromium.org/p/chromium/issues/detail?id=851873

**Comment by gabrielgrant** [comment]
FWIW it seems this issue is specific to high precision, and so is avoidable in three.js by forcing rendering with `mediump`. Is there any way to do that in AFrame?

**Comment by EerikKivistik** [comment]
@gabrielgrant I sent out test builds to some of our users who are having this issue and will report back if "mediump" solves the problem out in the wild.

**Comment by EerikKivistik** [comment]
@gabrielgrant So far, all the Android 7.0+ devices are rendering on "mediump" with the expected artifacts, but at least materials work. As for older devices running Android 5.0, it seems to make no difference, with everything still rendering completely black.

**Comment by gabrielgrant** [comment]
@EerikKivistik Looking back at [your original list of devices on which you'd seen the failure](https://github.com/mrdoob/three.js/issues/12679#issuecomment-375980801), they were all Android 7+ (except for the one Android 4.4 device, that i assumed was probably a red herring). Are there other devices you know of now that weren't listed there?

**Comment by EerikKivistik** [comment]
@gabrielgrant One example I recently discovered is XiaoMi MiPad 2 (Android version 5.1). The medium precision workaround has no effect there.  

As for the artifacts I will include a screenshot taken from an Android 7.0 device.
![screenshot_2018-06-14-22-55-57-765_com android browser](https://user-images.githubusercontent.com/5612934/41525396-8407ffd8-72e9-11e8-8524-38f456258801.png)

**Comment by gabrielgrant** [comment]
@EerikKivistik it seems the XiaoMi MiPad 2 has an Intel Atom X5-Z8500 (with an Intel HD Graphics GPU), whereas this issue appears to be specific to the Qualcomm Snapdragon 425, so it sounds like the issue(s) you're seeing with older devices is/are likely some other problem(s) that just happens to have the same symptom. May want to open another issue about those devices (probably against three.js rather than aframe). And in any case, probably best to continue this discussion in https://github.com/mrdoob/three.js/issues/14137

**Comment by bhouston** [comment]
WE here at Exocortex/Clara.io have more than 25 mobile devices.  Once we have a consistent test bed, we can test on each device and share a screenshot.  If everyone else does this we can likely pinpoint the issues and we can then work to fix them in a systematic way.

The trick is that each error should be reproducible at once in a single example without modifying any DAT-GUI settings.  Thus we only need a single screenshot for each device to know exactly where it stands with regards to bugs.

We do not even have to wait for this to be in the next release, we can use rawgit to view the example in either a PR state or once it is merged into master.

**Comment by ngokevin** [comment]
Duping with https://github.com/aframevr/aframe/issues/3971 ... there's a proposed PR ^

**Comment by EerikKivistik** [comment]
@gabrielgrant We finally went with "mediump" across all Android devices, as distinguishing between all the variations of devices was too complex for the size of our team to handle (and the availability to purchase the specific models was a problem). So far Android 7.0+ seems to be mostly working well. As for Android 5.0, it does produce some artifacts, but we will just have to live with that for the time being.
If anyone has an affected device, please do download 3dc.io on Android and give me feedback on whether or not this issue has also been solved for you. Thank you all for the debugging efforts!

**Comment by gabrielgrant** [comment]
@EerikKivistik That makes sense. Thanks to you too for all the device reports!

@dmarcos Thanks for finally getting around to tackling this at the aframe level, certainly didn't expect this issue to go quite so deep when I first started delving into it here and in mrdoob/three.js#14137

To confirm, it looks like the change to default precision was reverted in #3988 and then a `precision` prop was added in #3989

To confirm (and for others finding this issue), the recommendation is to manually set precision to `mediump` on the `a-scene`'s `renderer` component, right?

Would it make sense to just set this by default on devices with this GPU? I know doing device-specific fixes shouldn't really be a-frame's responsibility, but this is such a sharp edge that it seems like it might be worth working around.

If this is something you'd be open to, I could probably get a PR together that revives #3977 (making `mediump` the default) only for these devices

**Comment by dantepfer** [comment]
Just got an LG Aristo 3 to test this. Indeed setting `precision: low` or `precision: medium` fixes the issue, BUT I've found that if `antialias: true` is set, the issue returns (black shapes when using MeshStandardMaterial). So perhaps the recommendation should be not only to set `precision: medium`, but also to make sure that `antialias: false` is set (which is the default). 
Thanks much to all who contributed to this thread.
